File: code/src/main/scala/rpg/java/fast/SkillTrees.java
package rpg.java.fast;


import rpg.PlayerTraits;
import rpg.SkillTreeRepr;

final class Node {
    private final Node left;
    private final Node right;
    private final String name;
    private final boolean dex;
    private final boolean str;

    Node(boolean dex, boolean str, Node left, Node right, String name) {
        this.left = left;
        this.right = right;
        this.name = name;
        this.str = str;
        this.dex = dex;
    }

    int cost(PlayerTraits traits) {
        int cost = 0;
        if (dex) {
            cost = traits.dexCost();
        } else if (str) {
            cost = traits.strCost();
        } else {
            cost = traits.wisCost();
        }

        if (left != null) {
            cost = cost + left.cost(traits);
            if (right != null) {
                cost = cost + right.cost(traits);
            }
        }

        return cost;
    }

}

public class SkillTrees extends rpg.SkillTrees {


    @Override
    public SkillTreeRepr<?> archeryTree(PlayerTraits playerTraits) {
        Node tree = new Node(
                true, false,
                new Node(
                        false, false,
                        new Node(
                                true, false,
                                new Node(false, false, null, null, "aimed"),
                                new Node(false, false,
                                        new Node(true, false, null, null, "rapid"),
                                        null, null)
                                , null
                        ),
                        null, null),
                new Node(false, true,
                        new Node(false, true, null, null,"power"),
                        new Node(false, true,
                                new Node(true, false, null, null,"double"), null, null),
                        null
                ),
                null
        );


        return new SkillTreeRepr<Node>(tree) {
            @Override
            public int totalCost() {
                return tree.cost(playerTraits);
            }
        };
    }

    @Override
    public SkillTreeRepr<?> charismaTree(PlayerTraits traits) {
        Node tree = new Node(
          true, false,
          new Node(
            false, false,
            new Node(
              true, false,
              new Node(false, false, null, null, "aimed"),
              new Node(false, false,
                new Node(true, false, null, null, "rapid"),
                null, null)
              , null
            ),
            null, null),
          new Node(false, true,
            new Node(false, true, null, null,"power"),
            new Node(false, true,
              new Node(false, false, null, null,"double"), null, null),
            null
          ),
          null
        );


        return new SkillTreeRepr<Node>(tree) {
            @Override
            public int totalCost() {
                return tree.cost(traits);
            }
        };
    }
}

File: code/src/main/scala/rpg/SkillTrees.scala
package rpg

import org.openjdk.jmh.annotations.CompilerControl
import org.openjdk.jmh.annotations.CompilerControl.Mode.DONT_INLINE

class PlayerTraits(str : Int, dex : Int, wis : Int) {
  @CompilerControl(DONT_INLINE) def strCost: Int = str
  @CompilerControl(DONT_INLINE) def dexCost: Int = dex
  @CompilerControl(DONT_INLINE) def wisCost: Int = wis
}

abstract class SkillTreeRepr[T](val tree: T) {
  @CompilerControl(DONT_INLINE) def totalCost(): Int
}

abstract class SkillTrees {
  /** Dex --> Dex -> Wis --> Wis("aimed")
    *      \              \-> Wis -> Dex("rapid")
    *       \-> Str --> Str("power")
    *               \-> Str -> Dex("double") */
  def archeryTree(playerTraits: PlayerTraits): SkillTreeRepr[_]

  /** Wis -> Wis --> Wis("1")
    *            \-> Dex -> Dex --> Dex("2")
    *                           \-> Str -> Str --> Str("3")
    *                                          \-> Wis("4") */
  def charismaTree(traits: PlayerTraits): SkillTreeRepr[_]
}